### PR TITLE
Recurring payments support

### PR DESCRIPTION
**Pull request by PetrDlouhy**
This is implements recurring payments according to #64.
The code should be complete, but might require some polishing. I would appreciate any feedback with this.
I am currently implementing recurring payments with PayU through `django-payments` and `django-plans-payments`. The recurrence is induced by server (contrary to PayPal recurring payments), which means, that there is need to store recurring payment data as well as the card token.

### CODE REVIEW

**Comment by PetrDlouhy** [comment]
I consider this code ready for review.

File: plans/models.py [hunk 1]
```diff
@@ -349,6 +349,21 @@ def create_for_users_without_plan(cls):
         return userplans
 
 
+@python_2_unicode_compatible
+class RecurringUserPlan(models.Model):
```
Comment (Alir3z4) [inline plans/models.py:353]: Some docstring on this model would be great!

Explaining what it does and how it works and its role in the overall design

File: plans/tasks.py [hunk 2]
```diff
@@ -4,11 +4,33 @@
...
+from .signals import account_automatic_renewal
 
 
 User = get_user_model()
 logger = logging.getLogger('plans.tasks')
 
+
+@periodic_task(run_every=crontab(hour=0, minute=5))
+def autorenew_account():
+    logger.info('Started')
```
Comment (Alir3z4) [inline plans/tasks.py:16]: The logger could be more informative, at least containing the task name or what it's going to happen.

File: plans/management/commands/autorenew_accounts.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from django.core.management import BaseCommand
+from plans import tasks
+
+
+class Command(BaseCommand):
+    help = 'Autorenew accounts and with recurring payments'
```
Comment (Alir3z4) [inline plans/management/commands/autorenew_accounts.py:6]: This could be marked for translation ?

File: plans/models.py [hunk 4]
```diff
@@ -349,6 +349,21 @@ def create_for_users_without_plan(cls):
...
+    userplan = models.OneToOneField('UserPlan', on_delete=models.CASCADE, related_name='recurring')
```
Comment (Alir3z4) [inline plans/models.py:354]: Could be better to be `user_plan`

File: plans/models.py [hunk 5]
```diff
@@ -349,6 +349,21 @@ def create_for_users_without_plan(cls):
...
+    token = models.CharField(_('Recurring token'), max_length=255, default=None, null=True, blank=True)
```
Comment (Alir3z4) [inline plans/models.py:355]: verbose_name should be lowercase.

File: plans/models.py [hunk 6]
```diff
@@ -349,6 +349,21 @@ def create_for_users_without_plan(cls):
...
+    pricing = models.ForeignKey('Pricing', help_text=_('Recurring pricing'), default=None, null=True, blank=True, on_delete=models.CASCADE)
+    amount = models.DecimalField(
+        _('amount'), max_digits=7, decimal_places=2, db_index=True, null=True, blank=True)
+    tax = models.DecimalField(_('tax'), max_digits=4, decimal_places=2, db_index=True, null=True,
+                              blank=True)  # Tax=None is when tax is not applicable
+    currency = models.CharField(_('currency'), max_length=3, default='EUR')
```
Comment (Alir3z4) [inline plans/models.py:361]: Maybe we can turn the default currency into a settings variable ?

File: plans/models.py [hunk 7]
```diff
@@ -349,6 +349,21 @@ def create_for_users_without_plan(cls):
         return userplans
 
 
+@python_2_unicode_compatible
```
Comment (Alir3z4) [inline plans/models.py:352]: Lets remove the python 2 compat stuff as well

File: plans/tasks.py [hunk 8]
```diff
@@ -4,11 +4,33 @@
...
+    PLANS_AUTORENEW_BEFORE_DAYS = getattr(settings, 'PLANS_AUTORENEW_BEFORE_DAYS', 0)
+    PLANS_AUTORENEW_BEFORE_HOURS = getattr(settings, 'PLANS_AUTORENEW_BEFORE_HOURS', 0)
+
+    expired_accounts = User.objects.select_related('userplan').filter(
+        userplan__active=True,
+        userplan__expire__lt=datetime.date.today() + datetime.timedelta(days=PLANS_AUTORENEW_BEFORE_DAYS, hours=PLANS_AUTORENEW_BEFORE_HOURS),
+    ).exclude(userplan__expire=None)
+
```
Comment (Alir3z4) [inline plans/tasks.py:24]: Would be good to have a logger.info here saying something along the line of

```python
logger.info(f"{len(expired_accounts)} accounts to be renewed.")
```

File: plans/management/commands/autorenew_accounts.py (hunk 3 above)
Comment (PetrDlouhy) [reply plans/management/commands/autorenew_accounts.py:6]: This is management command - this text will be outputed only for system administrators. I don't think, it should be translated, and I don't even think, that translation would work on that.
And also all of the current management commands are untranslated.

Or do you execute management command from other place than command line?

File: plans/models.py (hunk 6 above)
Comment (PetrDlouhy) [reply plans/models.py:361]: Anywhere else in `django-plans` is `EUR` as default.
Maybe we could make additional commit, that changes that behaviour in entire system, but I think, there might raise some problems with migrations - new migrations would be generated on settings change.

Anyway, I don't think, that it is necessary to have any default for `RecurringUserPlan` at all. That object is created only with new order, so the currency should be always known. I will leave it with no default and test if it does not create any problems.

**Comment by PetrDlouhy** [comment]
@Alir3z4 I fixed most of your requests and commented the rest of them. I have also written some info about recurring payments to the docs.

I also realised, that we should not send reminders for recurring accounts. Although somebody might request other kind of reminders before recurring payment.

**Pull request merged**